python/ray/data/_internal/logical/rules/record_usage_tag.py

```diff
@@ -0,0 +1,117 @@
+from ray._private.usage.usage_lib import TagKey, record_extra_usage_tag
+from ray.data._internal.logical.interfaces import LogicalOperator, LogicalPlan, Rule
+from ray.data._internal.logical.operators.all_to_all_operator import (
+    Aggregate,
+    RandomizeBlocks,
+    RandomShuffle,
+    Repartition,
+    Sort,
+)
+from ray.data._internal.logical.operators.read_operator import Read
+from ray.data._internal.logical.operators.map_operator import (
+    Filter,
+    FlatMap,
+    MapBatches,
+    MapRows,
+    Write,
+)
+from ray.data.datasource.binary_datasource import BinaryDatasource
+from ray.data.datasource.csv_datasource import CSVDatasource
+from ray.data.datasource.datasource import Datasource
+from ray.data.datasource.image_datasource import ImageDatasource
+from ray.data.datasource.json_datasource import JSONDatasource
+from ray.data.datasource.mongo_datasource import MongoDatasource
+from ray.data.datasource.numpy_datasource import NumpyDatasource
+from ray.data.datasource.parquet_base_datasource import ParquetBaseDatasource
+from ray.data.datasource.parquet_datasource import ParquetDatasource
+from ray.data.datasource.text_datasource import TextDatasource
+from ray.data.datasource.tfrecords_datasource import TFRecordDatasource
+
+
+class RecordUsageTagRule(Rule):
```

Comment:
How does this compare to e.g. recording usage at planning time (logical->physical translation)? That looks a bit cleaner to me.

Reply:
I want to keep tracking the Dataset API usage, so think tag logical operator here (given logical operator is 1-to-1 mapping of API). If feeling awkward to put as optimization rule, I can move it before the logical optimization.

Reply:
Updated.